ratelimits/limiter_test.go

```diff
@@ -0,0 +1,305 @@
+package ratelimits
+
+import (
+	"testing"
+	"time"
+
+	"github.com/jmhodges/clock"
+	"github.com/letsencrypt/boulder/test"
+)
+
+const (
+	tenZeroZeroOne = "10.0.0.1"
+	tenZeroZeroTwo = "10.0.0.2"
+)
+
+// newTestLimiter makes a new limiter with the following configuration:
+//   - 'UsageRequestsPerIPv4Address' burst: 20 count: 20 period: 1s
+func newTestLimiter(t *testing.T) (*Limiter, clock.FakeClock) {
+	clk := clock.NewFake()
+	l, err := NewLimiter(clk, newInmem(), "testdata/defaults.yml", "")
+	test.AssertNotError(t, err, "should not error")
+	return l, clk
+}
+
+// newTestLimiterWithOverrides makes a new limiter with the following
+// configuration:
+//   - 'UsageRequestsPerIPv4Address' burst: 20 count: 20 period: 1s
+//   - 'UsageRequestsPerIPv4Address:10.0.0.2' burst: 40 count: 40 period: 1s
+func newTestLimiterWithOverrides(t *testing.T) (*Limiter, clock.FakeClock) {
+	clk := clock.NewFake()
+	l, err := NewLimiter(clk, newInmem(), "testdata/defaults.yml", "testdata/overrides.yml")
+	test.AssertNotError(t, err, "should not error")
+	return l, clk
+}
+
+func Test_Limiter_initialization_via_Check_and_Spend(t *testing.T) {
+	l, _ := newTestLimiter(t)
+
+	// Check on an empty bucket should initialize it and return the theoretical
+	// next state of that bucket if the cost were spent.
+	d, err := l.Check(UsageRequestsPerIPv4Address, tenZeroZeroOne, 1)
+	test.AssertNotError(t, err, "should not error")
+	test.Assert(t, d.Allowed, "should be allowed")
+	test.AssertEquals(t, d.Remaining, 19)
+	// Verify our RetryIn timing is correct. 1 second == 1000 milliseconds and
+	// 1000/20 = 50 milliseconds per request.
+	test.AssertEquals(t, d.ResetIn, time.Millisecond*50)
+	test.AssertEquals(t, d.RetryIn, time.Duration(0))
+
+	// However, that cost should not be spent yet, a 0 cost check should tell us
+	// that we actually have 20 remaining.
+	d, err = l.Check(UsageRequestsPerIPv4Address, tenZeroZeroOne, 0)
+	test.AssertNotError(t, err, "should not error")
+	test.Assert(t, d.Allowed, "should be allowed")
+	test.AssertEquals(t, d.Remaining, 20)
+	test.AssertEquals(t, d.ResetIn, time.Duration(0))
+	test.AssertEquals(t, d.RetryIn, time.Duration(0))
+
+	// Reset our bucket.
+	err = l.Reset(UsageRequestsPerIPv4Address, tenZeroZeroOne)
+	test.AssertNotError(t, err, "should not error")
+
+	// Similar to above, but we'll use Spend() instead of Check() to initialize
+	// the bucket. Spend should return the same result as Check.
```

Comment:
Here's an explicit check, though it uses two different limiters, that shows that Spend and Check return the same result.
```
	il, _ := newTestLimiter(t)
	jl, _ := newTestLimiter(t)
	i, err := il.Check(UsageRequestsPerIPv4Address, tenZeroZeroOne, 1)
	j, err := jl.Spend(UsageRequestsPerIPv4Address, tenZeroZeroOne, 1)
	test.AssertDeepEquals(t, i.Remaining, j.Remaining)
```

Reply:
Oooh, I like that this checks for parity.

Reply:
Added.